### ISSUE TITLE
Apparent concurrency issue with JsonConvert.DeserializeObject

### ISSUE THREAD

**Issue by mattklein999** [JamesNK/Newtonsoft.Json#469]
We're currently using `JsonConvert.SerializeObject` and `JsonConvert.DeserializeObject<T>` to send data between a client and server.

I created a tool that creates 10 clients, sends a command to 10 different servers, the servers serialize a response and send it back over the network, then the 10 clients deserialize the object on the local machine.

I'm running these 10 tasks concurrently in the thread pool, and about 20% of the time **all** of the 'JsonConvert.DeserializeObject<T>' calls fail with the following stack trace:

    Error: Index was outside the bounds of the array.
    at System.Collections.Generic.List1.Add(T item)
    at System.Collections.Generic.List1.System.Collections.IList.Add(Object item)
    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.PopulateList(IList list, JsonReader reader, JsonArrayContract contract, JsonProperty containerProperty, String id)
    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateList(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, Object existingValue, String id)
    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.SetPropertyValue(JsonProperty property, JsonConverter propertyConverter, JsonContainerContract containerContract, JsonProperty containerProperty, JsonReader reader, Object target)
    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.PopulateObject(Object newObject, JsonReader reader, JsonObjectContract contract, JsonProperty member, String id)
    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.SetPropertyValue(JsonProperty property, JsonConverter propertyConverter, JsonContainerContract containerContract, JsonProperty containerProperty, JsonReader reader, Object target)
    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.PopulateObject(Object newObject, JsonReader reader, JsonObjectContract contract, JsonProperty member, String id)
    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
    at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
    at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)
    at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value, JsonSerializerSettings settings)
    at MyClientCode()

`MyClientCode()` is using `DeserializeObject` like this:

    string json = GetServerResponse();
    return JsonConvert.DeserializeObject<ResponseObject>(json);

The `ResponseObject` is pretty big, and includes several composite objects. However, I can save the `json` and deserialize it properly with `DeserializeObject`, so I don't think the object structure is the problem.

Doing some research of the `List` error indicates that this occurs when trying to modify a `List` object concurrently.

I looked inside `JsonSerializerInternalReader` and noticed that it doesn't have any `lock`s or use the `System.Threading` namespace, but since it's an internal class I assume that thread synchronization is being handled by a public class.

Please let me know if I can provide any more information.

Thanks!

**Comment by JamesNK** [comment]
A new JsonSerializerInternalReader is created each time you deserialize an object. Each deserialization happens in its own state..

**Comment by mattklein999** [comment]
The cause being concurrency is just a theory. I suggest it because I can't reproduce the error when only creating one client, but can pretty consistently recreate it when creating 10 clients that all attempt to use `JsonConvert.DeserializeObject` at the same time.

Additionally, when a failure occurs, its _usually_ all 10 of the deserializations (from 10 different instances, with 10 different JSON strings, trying to deserialize the same type of object). But, sometimes, one will succeed and the other 9 will fail, and once I saw 4 succeed with the other 6 failing. When they fail, they always fail with the `Index was outside the bounds of the array` error. Also, if a failure occurs, I have to restart the program to get `JsonConvert.DeserializeObject` to work again. Any subsequent calls to it (with new JSON but the same deserialization type) throws the same exception.

To test if this may be caused by my code I saved a string of failed JSON response text to a file, then created a small console app that just loads the text from that file into memory and then fires off 30 tasks on the thread pool that deserialize the same text over and over. Usually, all 30 work, but sometimes all 30 fail, and rarely 1 will succeed and 29 will fail.

**Comment by JamesNK** [comment]
My guess is still you have multiple deserializers working over the same list instance. You either need to stop that or use a list that is thread safe.

**Comment by mattklein999** [comment]
Thanks for the suggestions James.

After digging deeper I found that you're right, we were using the same list object for multiple instances of the deserialized type. Specifically, the object looked something like this:

    class Obj {
        static List<string> _validSelections = new List<string>() { "One", "Two", "Three", "Four" };
        public IEnumerable<string> ValidSelections { get { return _validSelections; } }
        ... more ...
    }

The exception was being thrown on line 1261 of JsonSerializerInternalReader.cs when trying to add objects to the list concurrently.

After seeing how this is implemented in our code, I am going to get rid of the static backing since it wasn't providing us anything anyway. However, even if I lock around the deserializing, there is still an issue of having too many objects in each `obj.ValidSelections`.

For example, if I have 10 `Obj` objects, each with 4 values in `obj.ValidSelections`, then _every_ instance of `Obj` will end up with 40 elements in its `obj.ValidSelections` property. I'm not sure that this is really an issue with Json.NET or just bad design. If you feel that this is an actual issue, then I can create a new issue with more details.

Thanks again for your help and time!

---
I posted most of this thread to Stack Overflow just in case it helps others in the future:

http://stackoverflow.com/questions/28095244/jsonconvert-deserializeobject-index-was-outside-the-bounds-of-the-array/28095245#28095245
